distribution/push_v2.go

```diff
@@ -425,26 +422,34 @@ func (pd *v2PushDescriptor) uploadUsingSession(
 	diffID layer.DiffID,
 	layerUpload distribution.BlobWriter,
 ) (distribution.Descriptor, error) {
-	arch, err := pd.layer.TarStream()
-	if err != nil {
-		return distribution.Descriptor{}, xfer.DoNotRetry{Err: err}
-	}
-
-	// don't care if this fails; best effort
-	size, _ := pd.layer.DiffSize()
-
-	reader := progress.NewProgressReader(ioutils.NewCancelReadCloser(ctx, arch), progressOutput, size, pd.ID(), "Pushing")
-	compressedReader, compressionDone := compress(reader)
-	defer func() {
+	var (
+		reader io.ReadCloser
+	)
+
+	contentReader, err := pd.layer.Open()
+	size := pd.layer.Size()
+
+	reader = progress.NewProgressReader(ioutils.NewCancelReadCloser(ctx, contentReader), progressOutput, size, pd.ID(), "Pushing")
+
+	switch m := pd.layer.MediaType(); m {
+	case "application/vnd.docker.image.rootfs.diff.tar":
```

Comment:
`application/vnd.docker.container.image` is the one that went awry. This should have been `application/vnd.docker.image.config`, which is what is done in OCI.

Nothing is stopping these from being stored together. The mediatype isn't necessarily store with the object. The mediatype is part of the qualified reference, but not a property of the thing itself. For that to work, we'd have to hash the object with its type.